src/main/scala/com/fulcrumgenomics/fasta/SequenceInfos.scala

```diff
@@ -0,0 +1,171 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2020 Fulcrum Genomics
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.fasta
+
+import com.fulcrumgenomics.FgBioDef._
+import enumeratum.EnumEntry
+import htsjdk.samtools.{SAMSequenceDictionary, SAMSequenceRecord}
+
+
+/** Stores information about a single Sequence (ex. chromosome, contig)
+  *
+  * @param name the primary name of the sequence
+  * @param length the length of the sequence, or zero if unknown
+  * @param aliases the list of aliases for this sequence
+  * @param attributes attributes of this sequence
+  */
+case class SequenceInfo(name: String,
+                        length: Int,
+                        aliases: Seq[String] = Seq.empty,
+                        attributes: Map[String, Any] = Map.empty
+                       ) {
+
+  allNames.foreach { name => SAMSequenceRecord.validateSequenceName(name) }
+  require(length > 0, s"Length must be > 0 for '$name'")
+
+  /** All names, including aliases */
+  @inline final def allNames: Seq[String] = name +: aliases
+  @inline final def apply[T](key: String): T = this.attributes(key).asInstanceOf[T]
```

Comment:
I like these functions ... but does that imply we're going to do type-aware parsing of attributes when we read in dictionaries?  Otherwise maybe these should just be strings?

Reply:
I think this is tricky though because the header/dictionary format we're parsing doesn't have type tags on the attributes (e.g. doesn't allow `xi:i:273`) so when you're parsing from a file there is no way to know that "273" should be an `Int` and not a `String`.  And so if you allow for non-String types then you end up with problems like:

1. Do you get different types if constructed in memory (e.g. I put in an `Int` during construction) vs. when reading it back from disk?
2. Do you try and be "smart" when reading from disk and parse the attributes as Int/Float/String/List etc.?  But then you get inconsistent results.  What if I have one record with `xi:123` and another with `xi:foo`?  If I call `sequence[Int]("xi")` it will fail on one and if I call `sequence[String]("xi")` on the other it will fail because it is only _casting_ not doing type conversions.